File: Java/Hired Programming Practice/src/TreesAndGraphs/Q4.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package TreesAndGraphs;

import java.util.LinkedList;
import java.util.Queue;

/**
 *
 * @author marcsantiago
 */
//Given a binary tree, design an algorithm which creates a linked list 
//of all the nodes at each depth.
public class Q4 {
    //can be modified and added as a method within the BinaryTree defination
    public LinkedList treeToLinkedList(Node tree){
        if(tree != null){
            LinkedList<Integer> ll = new LinkedList<>();
            Queue<Node> q = new LinkedList<>();
            q.add(tree);
            while(!q.isEmpty()){
                Node node = q.remove();
                ll.add(node.data);
                if(node.leftPointer != null){
                    q.add(node.leftPointer);
                }
                if(node.rightPointer != null){
                    q.add(node.rightPointer);
                }
            }
            return ll;
        }
        return null;
    }
    
}

File: Python/Trees and Graphs/Q4.py
from BaseTreeImplementation import Node
# Given a binary tree, design an algorithm which creates a linked list
# of all the nodes at each depth.


def tree_to_linked_list(node):
    if node is not None:
        linked_list = []
        queue = []
        queue.insert(0, node)
        while queue:
            node = queue.pop()
            linked_list.append(node.data)
            if node.left is not None:
                queue.insert(0, node.left)

            if node.right is not None:
                queue.insert(0, node.right)


        return linked_list
    return None


def main():
    t = Node(10)
    t.left = Node(2)
    t.left.left = Node(1)
    t.left.right = Node(3)
    t.right = Node(20)
    t.right.right = Node(40)

    x = tree_to_linked_list(t)
    print x

if __name__ == '__main__':
    main()